sql/src/main/java/io/crate/ingestion/IngestionService.java

```diff
@@ -0,0 +1,163 @@
+
+    @Inject
+    public IngestionService(Settings settings,
+                            Schemas schemas,
+                            ClusterService clusterService,
+                            DDLClusterStateService ddlClusterStateService) {
+        super(settings);
+        this.schemas = schemas;
+        this.clusterService = clusterService;
+        ddlClusterStateService.addModifier(new IngestionDDLClusterStateModifier());
+    }
+
+    /**
+     * Register the provided @param implementation for ingest rule changes belonging to the provided @param sourceIdent.
+     */
+    public void registerImplementation(String sourceIdent, IngestionImplementation implementation) {
+        IngestionImplementation existingImplementation = implementations.put(sourceIdent, implementation);
+        if (existingImplementation != null) {
+            throw new IllegalArgumentException("There already exists an ingestion implementation registered for " +
+                                               sourceIdent);
+        }
+
+        Map<String, Set<IngestRule>> ingestionRules = getIngestRulesOrNull(clusterService.state().metaData());
+        if (ingestionRules == null) {
+            implementation.applyRules(Collections.emptySet());
+        } else {
+            implementation.applyRules(ingestionRules.getOrDefault(sourceIdent, Collections.emptySet()));
+        }
+    }
+
+    @VisibleForTesting
+    public void removeImplementationFor(String sourceIdent) {
+        implementations.remove(sourceIdent);
+    }
+
+    @Nullable
+    private Map<String, Set<IngestRule>> getIngestRulesOrNull(MetaData metaData) {
+        IngestRulesMetaData ingestRulesMetaData =
+            (IngestRulesMetaData) metaData.customs().get(IngestRulesMetaData.TYPE);
+
+        if (ingestRulesMetaData != null) {
+            return ingestRulesMetaData.getIngestRules();
+        }
+        return null;
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.metaDataChanged() == false) {
+            return;
+        }
+
+        Map<String, Set<IngestRule>> ingestionRules = getIngestionRulesIfChangedOrNull(event.state().metaData());
+        if(ingestionRules == null) {
+            return;
+        }
+        filterOutInvalidRules(ingestionRules);
+        for (Map.Entry<String, IngestionImplementation> entry : implementations.entrySet()) {
+            String sourceIdent = entry.getKey();
+            Set<IngestRule> rulesForSource = ingestionRules.getOrDefault(sourceIdent, Collections.emptySet());
+            entry.getValue().applyRules(rulesForSource);
+        }
+    }
+
+    private void filterOutInvalidRules(Map<String, Set<IngestRule>> ingestionRules) {
+        for (Set<IngestRule> sourceIngestRules : ingestionRules.values()) {
+            Iterator<IngestRule> rulesIterator = sourceIngestRules.iterator();
+            //noinspection Java8CollectionRemoveIf - would result in a lambda instance for every iteration
```

Comment:
I don't think we should disable it in general. If the lambda is not capturing it will be converted to a static function so it's not necessary always a bad conversion

Reply:
I think the false positive rate is high enough to justify disabling the warning completely. It's not as if the functional style is so superior vs. the imperative approach that it would hurt if we miss chances to use the functional style.